### PR TITLE
GeoJSON point clustering

### PR DESCRIPTION
**Pull request by mourner**
Clustering for GeoJSON point feature collections. Closes #1931. TODO:

- [x] make cluster maxzoom depend on GeoJSON source maxzoom
- [x] make cluster radius configurable
- [x] release a supercluster version and switch to it in package.json
- [x] JSDoc all API changes
- [x] revert the `site.js` changes
- [x] update mapbox-gl-test-suite https://github.com/mapbox/mapbox-gl-test-suite/pull/73
- [x] update mapbox-gl-style-spec https://github.com/mapbox/mapbox-gl-style-spec/pull/400
- [x] merge test-suite/spec
- [x] update deps to point to new test-suite/spec
- [x] squash into one commit

Per chat discussion, we're including supercluster in the core and will consider moving it out after we have a vision of how to do #929. Supercluster is a pretty tiny dependency (its biggest part is rbush which we already use) so won't affect size much.

cc @lucaswoj @ansis

### CODE REVIEW

**Comment by stepankuzmin** [comment]
aw yiss ✨

**Comment by mourner** [comment]
@lucaswoj @jfirebaugh this is pretty much ready to merge (after reverting `site.js`) with the only exception of a lack of tests — I'm not sure how to test this. Would it be OK to merge without tests and then rely on proper isolated tests in the supercluster repo?

**Comment by jfirebaugh** [comment]
Can you add one simple rendering example to mapbox-gl-test-suite, excluding from native [like this](https://github.com/mapbox/mapbox-gl-test-suite/blob/0fea16f7d1e0e0922f00da793e21a2420f6b8748/render-tests/video/default/style.json#L5) until clustering is ported there?

**Comment by mourner** [comment]
@jfirebaugh good idea, will do.

**Comment by lyzidiamond** [comment]
Tagging myself here to \o/ when this merges!

**Comment by lucaswoj** [comment]
Should we update mapbox-gl-style-spec to allow for GeoJSON cluster properties (`cluster`, `clusterRadius`, and `clusterMaxZoom`?

File: js/source/geojson_source.js [hunk 1]
```diff
@@ -42,10 +46,18 @@ function GeoJSONSource(options) {
...
-    this.geojsonVtOptions = { maxZoom: this.maxzoom };
+    this.geojsonVtOptions = {maxZoom: this.maxzoom};
     if (options.buffer !== undefined) this.geojsonVtOptions.buffer = options.buffer;
     if (options.tolerance !== undefined) this.geojsonVtOptions.tolerance = options.tolerance;
 
+    this.cluster = options.cluster || false;
+    this.superclusterOptions = {
+        maxZoom: Math.max(options.clusterMaxZoom, this.maxzoom - 1) || (this.maxzoom - 1),
+        extent: 4096,
+        radius: options.clusterRadius || 400,
```
Comment (lucaswoj) [inline js/source/geojson_source.js]: Why is the default here `400` but the [default in supercluster](https://github.com/mapbox/supercluster/blob/master/index.js#L20) `40`?

Comment (mourner) [inline js/source/geojson_source.js]: 400 relative to 4096 tile extent extent. Supercluster isn't tied to Mapbox Vector Tiles so I picked different defaults (40 of 512).

**Comment by mourner** [comment]
> Should we update mapbox-gl-style-spec to allow for GeoJSON cluster properties?

I'm not sure this belongs to the spec, in the same way geojson-vt options like simplification tolerance and tile buffer aren't. Spec describes purely visual representation of data, while clustering is data transformation (turning a bunch of points into one cluster point in a tile)..

**Comment by lucaswoj** [comment]
Won't https://github.com/mapbox/mapbox-gl-js/issues/1632 break clustered sources, then?

**Comment by mourner** [comment]
@lucaswoj should the validator complain about unrecognized properties in source options?  I'd say it shouldn't because it can be an extension point. If it should, I'd just add VT-related options as exceptions not documented in the spec.

**Comment by lucaswoj** [comment]
@mourner Ah! You're right. We only restrict top-level properties on the root and layers (https://github.com/mapbox/mapbox-gl-style-spec/issues/325). I'm :+1: on this PR. Lets try to get one more set of eyes on it before merging, though. @ansis @jfirebaugh?

**Comment by mourner** [comment]
The only thing I'm not sure about is how/where to document the clusters it generates — maybe in description of `cluster: true` option of `GeoJSONSource`? It creates tiles with point features that have the following properties: `cluster: true`, `numPoints` (number of clustered points) and `numPointsH` (human-readable, e.g. "200k").

**Pull request merged**